src/pkg/k8s/namespace.go

```diff
@@ -45,19 +45,20 @@ func (k *K8s) DeleteNamespace(ctx context.Context, name string) error {
 	// Attempt to delete the namespace immediately
 	gracePeriod := int64(0)
 	err := k.Clientset.CoreV1().Namespaces().Delete(ctx, name, metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod})
-	// If an error besides "not found" is returned, return it
 	if err != nil && !errors.IsNotFound(err) {
 		return err
 	}
 
-	// Indefinitely wait for the namespace to be deleted, use context.WithTimeout to limit this
 	for {
-		// Keep checking for the namespace to be deleted
-		_, err := k.Clientset.CoreV1().Namespaces().Get(ctx, name, metav1.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(1 * time.Second):
```

Comment:
Yeah good point about the initial wait. We should be using `time.After()` rather than `time.Sleep()` though because it is non-blocking. We can change the order of operations so that there is no pause before executing the logic we want:

```go
for {
      _, err := k.Clientset.CoreV1().Namespaces().Get(ctx, name, metav1.GetOptions{})
      if errors.IsNotFound(err) {
          return nil
      }
      select {
      case <-ctx.Done():
          return ctx.Err()
      case <-time.After(1 * time.Second):
      }
}
```

Reply:
A relatively rare but possible case is the context getting cancelled by the parent function instead of by timeout. In that case, ctx could be done before the first loop and the caller would expect ctx.Err() returned rather than a successful attempt. I think I've found the pattern we're looking for actually

```
func main() {
	start := time.Now()
	instantButCancel()
	fmt.Println(time.Since(start))
	instantCancel()
}

func instantButCancel() {
	ctx, _ := context.WithTimeout(context.TODO(), 500 * time.Millisecond)
	timer := time.NewTimer(0) 
	for {
		select {
		case <-ctx.Done():
			return 
		case <-timer.C:
			fmt.Println("called by instantNoWait")
			timer.Reset(1 * time.Second)
		}
	}
}

func instantCancel() {
	ctx, cancel := context.WithCancel(context.TODO())
	cancel()
	timer := time.NewTimer(0) 
	for {
		select {
		case <-ctx.Done():
			fmt.Println("in done block instant cancel")
			return 
		case <-timer.C:
			fmt.Println("not called by instantCancel")
			timer.Reset(1 * time.Second)
		}
	}
}
```

Running this yields which I believe is everything we want
```
called by instantNoWait
500.650383ms
in done block instant cancel
```

Reply:
That seems reasonable. Where would the actual logic get executed with this pattern?

Reply:
Under timer.C like this, goes in instantly first loop, all loops after that have a 1 second delay and are nonblocking on ctx
```
	timer := time.NewTimer(0) 
	for {
		select {
		case <-ctx.Done():
			fmt.Println("in done block instant cancel")
			return 
		case <-timer.C:
			fmt.Println("do the logic here")
			timer.Reset(1 * time.Second)
		}
	}
```